pkg/ldbwriter/ldb_callback_writer.go

```diff
@@ -3,33 +3,105 @@ package ldbwriter
 import (
 	"context"
 	"database/sql"
-
 	"github.com/segmentio/ctlstore/pkg/schema"
 	"github.com/segmentio/ctlstore/pkg/sqlite"
 	"github.com/segmentio/events/v2"
+	"github.com/segmentio/stats/v4"
 )
 
 // CallbackWriter is an LDBWriter that delegates to another
 // writer and then, upon a successful write, executes N callbacks.
 type CallbackWriter struct {
-	DB           *sql.DB
-	Delegate     LDBWriter
-	Callbacks    []LDBWriteCallback
+	DB        *sql.DB
+	Delegate  LDBWriter
+	Callbacks []LDBWriteCallback
+	// Buffer between SQLite preupdate Hook and this code
 	ChangeBuffer *sqlite.SQLChangeBuffer
+	// Accumulated changes across multiple ApplyDMLStatement calls
+	transactionChanges []sqlite.SQLiteWatchChange
+}
+
+func (w *CallbackWriter) inTransaction() bool {
+	return w.transactionChanges != nil
+}
+
+func (w *CallbackWriter) beginTransaction(ledgerSequence schema.DMLSequence) {
+	if len(w.transactionChanges) > 0 {
+		// This should never happen, but just in case...
+		stats.Add("ldb_changes_abandoned", len(w.transactionChanges))
+		events.Log("error: abandoned %{count}d changes from incomplete transaction, current statement's ledger sequence: %{sequence}d",
+			len(w.transactionChanges), ledgerSequence)
+	}
+	w.transactionChanges = make([]sqlite.SQLiteWatchChange, 0)
+	// TODO: Figure out if we wanna use a gauge or a counter here
+	stats.Set("ldb_changes_accumulated", 0)
+}
+
+// Transaction done! Return the accumulated changes including the latest ones
+func (w *CallbackWriter) endTransaction(changes *[]sqlite.SQLiteWatchChange) {
```

Comment:
does changes really need to be a pointer? Its also a bit odd that you append to the changes slice instead of the w.transactionChanges slice.